bolt-design-system/bolt
# Improve UIKit Rendering Performance + Speed Up CI Installs

## Jira
http://vjira2:8080/browse/BDS-1557

## Summary
DevOps related updates to upgrade Bolt’s Puppeteer instances to now use `Puppeteer Core` + launch using an instance of Chrome _already installed_; significantly speeding up / slimming down local dev installs + shaving nearly a minute off of our Travis CI build!

Also includes a handful of UIKit visual improvements + a ton of IE-specific bug fixes that popped up when testing out these updates.

**Before:**
![image](https://user-images.githubusercontent.com/1617209/60976469-bd6c6780-a2fb-11e9-9b93-c33ea3fabaf2.png)

**After:**
![image](https://user-images.githubusercontent.com/1617209/60976541-decd5380-a2fb-11e9-8dc8-12cbb56c38fe.png)


## Details
- This PR updates dependencies and configuration logic for Bolt’s Puppeteer instanced to now use the much lighter weight Puppeteer Core instead (which doesn’t automatically download Chrome when installing). This includes: 
  - Web component-specific Jest tests
  - Pre-rendering Pattern Lab’s Web Component-based UI
  - Generating and inlining Critical CSS for Pattern Lab (and soon, the rest of the docs site)

- Updated NPM dependencies for the root monorepo `package.json`, in addition to updating dependencies in Pattern Lab's UI (UIKit Workshop), `@bolt/core`, and Bolt's shared Babel-specific config
- Fixed the double CLI output issue when Pattern Lab's UIKit assets are compiled
- Fixed a few misc eslint / prettier related issues flagged after updating dependencies
- Added / updated several missing polyfills required to get IE 11 working as expected
- Removed Prettier from getting used to format code in the Pattern Lab code viewer due to major performance issues encountered (now using a lighter weight alternative)
- Reverted Pattern Lab's iframe to no longer automatically resize when the height changes. This addresses side-effects we've been seeing with sticky / fixed position UI (like modals) + helps improve the already heavy performance load when navigating between pages
- Updated Pattern Lab's iframe scrollbar / iframe resizer handle to be easier to use while keeping out of the way 
- Improvements to the no-JS / pre-rendered version of the docs site -- the nav + iframe are nav both scrollable / usable!

### Prerendering Before vs After Updates
![image](https://user-images.githubusercontent.com/1617209/61079712-647efb00-a3f1-11e9-97ee-3ab0b9471f0d.png)


### What's Up With The New UIKit-specific Packages?
Our modernized Pattern Lab UI had been using two off the shelf Webpack plugins (`critical-css-webpack-plugin` and `prerender-spa-plugin`) to handle the pre-rendering and Critical CSS-related work. Unfortunately, these two plugins (and the `prerenderer` dependency used for pre-rendering) were also using the heavier version of Puppeteer. 

This, combined with a lack of recent upstream updates, meant that configuring these Puppeteer instances to use Puppeteer Core + reference the path to a local instance of Chrome installed was incredibly difficult. 

For at least the time being, these 3 dependencies have been pulled into the Bolt monorepo to apply the necessary optimizations.
> These updates could ultimately get PR’d back to the original projects, however given the lack of recent package updates, my gut tells me that a `patch-package`-approach would likely wind up being much more permanent than I’m comfortable with.

## How to test
- Confirm Travis build working as expected
- Double-check the updates locally to confirm everything here is good to go
- Check out the minor UI updates in [this PR's staged version of Pattern Lab ](https://feature-uikit-pre-rendering-updates.boltdesignsystem.com/pattern-lab/?p=components-overview)
